### PR TITLE
feat(server): short-hand routers and spread support

### LINKED ISSUE #3630: feat: add shorthand `{}` for sub-routers
### Describe the feature you'd like to request

I don't wanna write `router({ })` for my sub-routers

### Describe the solution you'd like to see

These should be equivalent

```ts
const appRouter = router({
  foo: router({
    bar: publicProcedure.query(() => '...'),
  })
})
```

```ts
const appRouter = router({
  foo: {
    bar: publicProcedure.query(() => '...'),
  }
})
```

### Describe alternate solutions

Not doing it

### Additional information

If it creates tech debt, don't do it.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

<sub>[TRP-49](https://linear.app/trpc/issue/TRP-49/feat-add-shorthand-for-sub-routers)</sub>

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/3630">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/3630/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by KATT**
- Allow for creating routers that are just objects
- Will simplify `.d.ts`-files for people who use `satisfies TRPCRouterRecord`

Closes #3630

Could help make #4129 easier

### How it looks like

This will allow you to define routers like this:

```ts

const post = {
  byId: t.procedure.query(() => '..'),
} satisfies RouterRecord; // <--  ✨ using satisfies for sub-routers could(?) be less heavy for the compiler and `.d.ts`-files



const appRouter = t.router({
  post,
  // ✨ short-cut routers by doing a `{}`
  deeply: {
    nested: {
      greeting: t.query(() => '...'),
    },
  },
});

```

### CODE REVIEW

File: packages/client/src/createTRPCClient.ts [hunk 1]
```diff
@@ -78,12 +78,14 @@ type DecorateProcedure<
...
 type DecoratedProcedureRecord<
   TRouter extends AnyRouter,
-  TProcedures extends ProcedureRouterRecord,
+  TRecord extends RouterRecord,
 > = {
-  [TKey in keyof TProcedures]: TProcedures[TKey] extends AnyRouter
-    ? DecoratedProcedureRecord<TRouter, TProcedures[TKey]['_def']['record']>
-    : TProcedures[TKey] extends AnyProcedure
-    ? DecorateProcedure<TRouter['_def']['_config']['$types'], TProcedures[TKey]>
+  [TKey in keyof TRecord]: TRecord[TKey] extends infer $Value
```
Comment (juliusmarminge) [inline packages/client/src/createTRPCClient.ts:83]: why do you need this `extends infer` here?

File: packages/tests/server/mergeRouters.test.ts [hunk 2]
```diff
@@ -17,6 +18,26 @@ test('mergeRouters', async () => {
...
+    foo: t.procedure.query(() => 'foo'),
+  } satisfies TRPCRouterRecord;
+  const router2 = {
+    bar: t.procedure.query(() => 'bar'),
+  } satisfies TRPCRouterRecord;
+
+  const merged = t.router({
+    ...router1,
+    ...router2,
+  });
```
Comment (juliusmarminge) [inline packages/tests/server/mergeRouters.test.ts:34]: niceee - can we deprecate this monstosity then?

https://github.com/trpc/trpc/blob/next/packages/server/src/unstable-core-do-not-import/router.ts#L272-L309

or maybe just make it an alias?

**Review by juliusmarminge — COMMENTED** [review]
Great work, only nitpick I'd point out is that there's a lot of these:

```ts
  TRouter['_def']['_config']['$types'],
  TRouter['_def']['record']
```

don't we have inference helpers for those?

```ts
  inferRootTypes<TRouter>
  ...
```

?

File: packages/client/src/createTRPCClient.ts (hunk 1 above)
Comment (KATT) [reply packages/client/src/createTRPCClient.ts:83]: I just think it makes the "looping" clearer

Comment (KATT) [reply packages/client/src/createTRPCClient.ts:83]: Only need to grab `TRecord[Key]` once, it's like a declaring a variable

Comment (juliusmarminge) [reply packages/client/src/createTRPCClient.ts:83]: > Only need to grab `TRecord[Key]` once, it's like a declaring a variable

aah i see, how is the compiler performance comparatively?

**Comment by KATT** [comment]
> Great work, only nitpick I'd point out is that there's a lot of these:
> 
> ```ts
>   TRouter['_def']['_config']['$types'],
>   TRouter['_def']['record']
> ```
> 
> don't we have inference helpers for those?
> 
> ```ts
>   inferRootTypes<TRouter>
>   ...
> ```
> 
> ?

I'd be fine to change that, I'd also be happy if we changed the router so the grabbing would just be e.g. 

- `TRouter['_def']['$types']`
- `TRouter['_def']['_config']['$types']`
- `TRouter[RootTypesSymbol]`
- `TRouter['_config']['$types']`

I think that sort of approach would be even nicer than using a helper - it feels like an unnecessary abstraction

File: packages/client/src/createTRPCClient.ts (hunk 1 above)
Comment (KATT) [reply packages/client/src/createTRPCClient.ts:83]: No idea tbh 🙈

File: packages/tests/server/mergeRouters.test.ts (hunk 2 above)
Comment (KATT) [reply packages/tests/server/mergeRouters.test.ts:34]: yes that'd be nice

Comment (KATT) [reply packages/tests/server/mergeRouters.test.ts:34]: we could alias it

**Review by juliusmarminge — APPROVED** [review]
i see that you can't spread actual routers:

```ts
// router/post.ts
import { createTRPCRouter, protectedProcedure, publicProcedure } from "../trpc";

export const authRouter = createTRPCRouter({
  getSession: publicProcedure.query(({ ctx }) => {
    return ctx.session;
  }),
  getSecretMessage: protectedProcedure.query(() => {
    // testing type validation of overridden next-auth Session in @acme/auth package
    return "you can see this secret message!";
  }),
});

// root.ts
import { authRouter } from "./router/auth";
import { postRouter } from "./router/post";
import { createTRPCRouter } from "./trpc";

export const appRouter = createTRPCRouter({
  auth: authRouter,
  post: postRouter,
  ...authRouter, // <-- doesn't work
});

// export type definition of API
export type AppRouter = typeof appRouter;
```

how do you make the typesystem / proxy understand spread? if we to alias mergeRouters to this implementation we should support spreading routers and not just records?

CleanShot 2024-01-30 at 08 42 50@2x

**Comment by KATT** [comment]
> i see that you can't spread actual routers:
> 
> ```ts
> // router/post.ts
> import { createTRPCRouter, protectedProcedure, publicProcedure } from "../trpc";
> 
> export const authRouter = createTRPCRouter({
>   getSession: publicProcedure.query(({ ctx }) => {
>     return ctx.session;
>   }),
>   getSecretMessage: protectedProcedure.query(() => {
>     // testing type validation of overridden next-auth Session in @acme/auth package
>     return "you can see this secret message!";
>   }),
> });
> 
> // root.ts
> import { authRouter } from "./router/auth";
> import { postRouter } from "./router/post";
> import { createTRPCRouter } from "./trpc";
> 
> export const appRouter = createTRPCRouter({
>   auth: authRouter,
>   post: postRouter,
>   ...authRouter, // <-- doesn't work
> });
> 
> // export type definition of API
> export type AppRouter = typeof appRouter;
> ```
> 
> how do you make the typesystem / proxy understand spread? if we to alias mergeRouters to this implementation we should support spreading routers and not just records?

No idea, but I think most people use sub-routers with the other way, so might be a non-issue

**Pull request merged**